aws-serverless-java-container-core/src/main/java/com/amazonaws/serverless/proxy/internal/servlet/AwsProxyHttpServletRequest.java

```diff
@@ -345,15 +354,35 @@ public Enumeration<String> getParameterNames() {
         if (request.getMultiValueQueryStringParameters() == null) {
             return Collections.enumeration(formParameterNames);
         }
-        return Collections.enumeration(Stream.concat(formParameterNames.stream(),
-                request.getMultiValueQueryStringParameters().keySet().stream()).collect(Collectors.toSet()));
+
+        Set<String> paramNames = request.getMultiValueQueryStringParameters().keySet();
+        if (request.getRequestSource() == RequestSource.ALB) {
+        	paramNames = paramNames.stream().map(AwsProxyHttpServletRequest::decodeValueIfEncoded).collect(Collectors.toSet());
+        }
+		
+        return Collections.enumeration(
+				Stream.concat(formParameterNames.stream(), paramNames.stream())
+				.collect(Collectors.toSet()));
     }
 
 
     @Override
     @SuppressFBWarnings("PZLA_PREFER_ZERO_LENGTH_ARRAYS") // suppressing this as according to the specs we should be returning null here if we can't find params
     public String[] getParameterValues(String s) {
+    	
+    	// decode key if ALB
+    	if (request.getRequestSource() == RequestSource.ALB) {
+    		s = decodeValueIfEncoded(s);
+    	}
+    	
+    	// TODO lots of back and forth arrays and lists here, sort it out!
         List<String> values = new ArrayList<>(Arrays.asList(getQueryParamValues(request.getMultiValueQueryStringParameters(), s, config.isQueryStringCaseSensitive())));
+        // List<String> values = getQueryParamValuesAsList(request.getMultiValueQueryStringParameters(), s, config.isQueryStringCaseSensitive());
```

Comment:
I imagine this comment shouldn't be here

Reply:
I also had a todo around cleaning up the conversions between arrays and lists happening here and the getQueryParamValues() method. This should be a little better with some extra null safety now and the commented code has been removed.